types/int_test.go

```diff
@@ -109,3 +110,211 @@ func TestUint(t *testing.T) {
+		ni = NewInt(n)
+
+		require.Equal(t, ni, NewInt(ni.Int64()))
+		require.Equal(t, ni, NewIntFromBigInt(ni.BigInt()))
+		nifromstr, ok = NewIntFromString(ni.String())
+		require.True(t, ok)
+		require.Equal(t, ni, nifromstr)
+		require.Equal(t, ni, NewInt(ni.BigInt().Int64()))
+		require.Equal(t, ni, NewIntWithDecimal(n, 0))
+
+		// Test purity
+		require.Equal(t, NewInt(n), ni)
+
+		// Test homomorphicity
+		n1 := int64(rand.Int31())
+		ni1 := NewInt(n1)
+		n2 := int64(rand.Int31())
+		ni2 := NewInt(n2)
+
+		testArithInt(t, n2, n1+n2, ni1.Add, ni1.AddRaw)
+		testArithInt(t, n2, n1-n2, ni1.Sub, ni1.SubRaw)
+		testArithInt(t, n2, n1*n2, ni1.Mul, ni1.MulRaw)
+		testArithInt(t, n2, n1/n2, ni1.Div, ni1.DivRaw)
+		require.Equal(t, func(i1 int64, i2 int64) int64 {
+			if i1 < i2 {
+				return i1
+			}
+			return i2
+		}(n1, n2), MinInt(ni1, ni2).Int64())
+		require.Equal(t, -n1, ni1.Neg().Int64())
+
+		require.Equal(t, n1 == n2, ni1.Equal(ni2))
+		require.Equal(t, n1 > n2, ni1.GT(ni2))
+		require.Equal(t, n1 < n2, ni1.LT(ni2))
+	}
+}
+
+func testArithUint(t *testing.T, n2 uint64, res uint64, fn func(Uint) Uint, fnraw func(uint64) Uint) {
+	require.Equal(t, res, fn(NewUint(n2)).Uint64())
+	require.Equal(t, res, fnraw(n2).Uint64())
+}
+
+func TestConsistencyUint(t *testing.T) {
+	for i := 0; i < 20; i++ {
+
+		// Test identity: uint64 -> Uint/big.Int/string -> uint64
+		n := uint64(rand.Uint32())
+		ni := NewUint(n)
+
+		require.Equal(t, n, ni.Uint64())
+		require.Equal(t, n, ni.BigInt().Uint64())
+		nifromstr, ok := NewUintFromString(strconv.FormatUint(n, 10))
+		require.True(t, ok)
+		require.Equal(t, n, nifromstr.Uint64())
+		require.Equal(t, n, NewUintFromBigInt(big.NewInt(int64(n))).Uint64())
+		require.Equal(t, n, NewUintWithDecimal(n, 0).Uint64())
+
+		// Test purity
+		require.Equal(t, NewUint(n), ni)
+
+		// Test identity: Uint -> uint64/big.Int/string -> Uint
+		ni = NewUint(n)
+
+		require.Equal(t, ni, NewUint(ni.Uint64()))
+		require.Equal(t, ni, NewUintFromBigInt(ni.BigInt()))
+		nifromstr, ok = NewUintFromString(ni.String())
+		require.True(t, ok)
+		require.Equal(t, ni, nifromstr)
+		require.Equal(t, ni, NewUint(ni.BigInt().Uint64()))
+		require.Equal(t, ni, NewUintWithDecimal(n, 0))
+
+		// Test purity
+		require.Equal(t, NewUint(n), ni)
+
+		// Test homomorphicity
+		n1 := uint64(rand.Uint32())
+		ni1 := NewUint(n1)
+		n2 := uint64(rand.Uint32())
+		ni2 := NewUint(n2)
+
+		testArithUint(t, n2, n1+n2, ni1.Add, ni1.AddRaw)
+		if n1 > n2 {
+			testArithUint(t, n2, n1-n2, ni1.Sub, ni1.SubRaw)
+		}
+		testArithUint(t, n2, n1*n2, ni1.Mul, ni1.MulRaw)
+		testArithUint(t, n2, n1/n2, ni1.Div, ni1.DivRaw)
+		require.Equal(t, func(i1 uint64, i2 uint64) uint64 {
+			if i1 < i2 {
+				return i1
+			}
+			return i2
+		}(n1, n2), MinUint(ni1, ni2).Uint64())
+
+		require.Equal(t, n1 == n2, ni1.Equal(ni2))
+		require.Equal(t, n1 > n2, ni1.GT(ni2))
+		require.Equal(t, n1 < n2, ni1.LT(ni2))
+	}
+}
+
+func TestImmutability(t *testing.T) {
```

Comment:
I don't think the way this immutability tests is written super extendable / table driven. I think perhaps we should define a struct called `operation` which contains an `operation code`, another uint to act upon. (Or alternatively this struct just has function for applying, as you've currently done). Then we can adapt `testArithUint` to apply that op-code (instead of the function directly). Then a single entry of our test-table is going to be a start number, and a list of `operations`. This is for easy extendability

We don't needs to have these operations list out the expected value at that point. (Feel free to ignore this point, I think either way makes equal sense) Instead what we can do is when processing a table entry, create a slice of `len(operations) + 2` Ints. Then populate this list with copies of the original list. (Assumes we have a method for copying that won't tie the two pointers together, if not we should be able to make one. Also I imagine this a fair assumption) At the end make sure all of the original values are still equal to this list of values.